File: analyze/rowcoach/geometry.py
"""Geometry helpers: angles, side detection, and trajectory smoothing.

All inputs are landmark arrays from pose.py with image coordinates (y DOWN).
Angle helpers convert to a math-friendly frame (y UP) internally so that
"degrees from vertical" reads the way a human expects.
"""

from __future__ import annotations

from typing import Literal

import numpy as np
from scipy.signal import savgol_filter

from . import pose
from .config import SMOOTH_POLYORDER, SMOOTH_WINDOW

Facing = Literal["left", "right"]


def smooth_landmarks(landmarks: np.ndarray,
                     window: int = SMOOTH_WINDOW,
                     polyorder: int = SMOOTH_POLYORDER) -> np.ndarray:
    """Savitzky-Golay smoothing of x,y trajectories over time (axis 0).

    Visibility channel is left untouched. No-op if the clip is shorter than the
    smoothing window.
    """
    n = landmarks.shape[0]
    if n < window or window < 3:
        return landmarks.copy()
    win = window if window % 2 == 1 else window + 1
    win = min(win, n if n % 2 == 1 else n - 1)
    out = landmarks.copy()
    out[:, :, 0] = savgol_filter(landmarks[:, :, 0], win, polyorder, axis=0)
    out[:, :, 1] = savgol_filter(landmarks[:, :, 1], win, polyorder, axis=0)
    return out


def detect_facing(landmarks: np.ndarray) -> Facing:
    """Decide whether the rower faces left or right in the frame.

    Heuristic: on an erg the handle/wrists travel toward the flywheel in front of
    the rower. The nose sits ahead of the hips horizontally in the facing
    direction. Average nose-vs-hip x over high-confidence frames.
    """
    nose_x = landmarks[:, pose.NOSE, 0]
    hip_x = (landmarks[:, pose.L_HIP, 0] + landmarks[:, pose.R_HIP, 0]) / 2.0
    # weight by hip visibility to ignore junk frames
    w = (landmarks[:, pose.L_HIP, 2] + landmarks[:, pose.R_HIP, 2]) / 2.0
    if w.sum() <= 0:
        return "right"
    delta = float(np.average(nose_x - hip_x, weights=w))
    # nose to the right of hips (larger x) -> facing right
    return "right" if delta >= 0 else "left"


def side_indices(facing: Facing) -> dict[str, int]:
    """Landmark indices for the camera-facing side of the body."""
    if facing == "left":
        return {
            "shoulder": pose.L_SHOULDER, "elbow": pose.L_ELBOW,
            "wrist": pose.L_WRIST, "hip": pose.L_HIP,
            "knee": pose.L_KNEE, "ankle": pose.L_ANKLE,
        }
    return {
        "shoulder": pose.R_SHOULDER, "elbow": pose.R_ELBOW,
        "wrist": pose.R_WRIST, "hip": pose.R_HIP,
        "knee": pose.R_KNEE, "ankle": pose.R_ANKLE,
    }


def _to_xy_up(p: np.ndarray) -> np.ndarray:
    """Convert an [..., (x,y,...)] point from image coords (y down) to y-up."""
    q = p[..., :2].astype(np.float64).copy()
    q[..., 1] = -q[..., 1]
    return q


def joint_angle(a: np.ndarray, b: np.ndarray, c: np.ndarray) -> float:
    """Interior angle (degrees) at vertex b formed by points a-b-c."""
    a, b, c = _to_xy_up(a), _to_xy_up(b), _to_xy_up(c)
    ba = a - b
    bc = c - b
    denom = (np.linalg.norm(ba) * np.linalg.norm(bc))
    if denom == 0:
        return float("nan")
    cosang = np.clip(np.dot(ba, bc) / denom, -1.0, 1.0)
    return float(np.degrees(np.arccos(cosang)))


def angle_from_vertical(a: np.ndarray, b: np.ndarray) -> float:
    """Angle (degrees) of segment a->b away from vertical (0 = vertical).

    Always non-negative; pair with a sign helper when lean direction matters.
    """
    a, b = _to_xy_up(a), _to_xy_up(b)
    v = b - a
    if np.linalg.norm(v) == 0:
        return float("nan")
    # angle vs the +y (up) axis
    cosang = np.clip(v[1] / np.linalg.norm(v), -1.0, 1.0)
    return float(np.degrees(np.arccos(abs(cosang))))


def signed_lean_from_vertical(hip: np.ndarray, shoulder: np.ndarray,
                              facing: Facing) -> float:
    """Torso lean from vertical in degrees, signed so that:

      positive  = leaning toward the catch (forward / shoulders ahead of hips)
      negative  = leaning toward the finish (back / shoulders behind hips)

    'Forward' is the facing direction. Used for both catch lean and finish layback.
    """
    h, s = _to_xy_up(hip), _to_xy_up(shoulder)
    v = s - h
    if np.linalg.norm(v) == 0:
        return float("nan")
    mag = np.degrees(np.arccos(np.clip(v[1] / np.linalg.norm(v), -1.0, 1.0)))
    # horizontal component: shoulder ahead of hip in facing direction => forward
    dx = v[0]
    forward = dx >= 0 if facing == "right" else dx <= 0
    return float(mag if forward else -mag)

File: live/src/geometry.ts
// Geometry helpers mirroring the Python Tier 2 conventions, but for MoveNet's
// 17 COCO keypoints in pixel coords (y DOWN). Angles convert to y-up internally.

export interface KP {
  x: number;
  y: number;
  score: number;
}

// MoveNet / COCO-17 keypoint indices
export const KEYPOINT = {
  nose: 0,
  leftShoulder: 5,
  rightShoulder: 6,
  leftElbow: 7,
  rightElbow: 8,
  leftWrist: 9,
  rightWrist: 10,
  leftHip: 11,
  rightHip: 12,
  leftKnee: 13,
  rightKnee: 14,
  leftAnkle: 15,
  rightAnkle: 16,
} as const;

export type Facing = "left" | "right";

export interface SideIdx {
  shoulder: number;
  elbow: number;
  wrist: number;
  hip: number;
  knee: number;
  ankle: number;
}

export function sideIndices(facing: Facing): SideIdx {
  return facing === "left"
    ? {
        shoulder: KEYPOINT.leftShoulder,
        elbow: KEYPOINT.leftElbow,
        wrist: KEYPOINT.leftWrist,
        hip: KEYPOINT.leftHip,
        knee: KEYPOINT.leftKnee,
        ankle: KEYPOINT.leftAnkle,
      }
    : {
        shoulder: KEYPOINT.rightShoulder,
        elbow: KEYPOINT.rightElbow,
        wrist: KEYPOINT.rightWrist,
        hip: KEYPOINT.rightHip,
        knee: KEYPOINT.rightKnee,
        ankle: KEYPOINT.rightAnkle,
      };
}

const up = (p: KP): [number, number] => [p.x, -p.y];

export function jointAngle(a: KP, b: KP, c: KP): number {
  const [ax, ay] = up(a), [bx, by] = up(b), [cx, cy] = up(c);
  const bax = ax - bx, bay = ay - by;
  const bcx = cx - bx, bcy = cy - by;
  const denom = Math.hypot(bax, bay) * Math.hypot(bcx, bcy);
  if (denom === 0) return NaN;
  const cos = Math.max(-1, Math.min(1, (bax * bcx + bay * bcy) / denom));
  return (Math.acos(cos) * 180) / Math.PI;
}

// Angle of segment a->b away from vertical (0 = vertical), always >= 0.
export function angleFromVertical(a: KP, b: KP): number {
  const [ax, ay] = up(a), [bx, by] = up(b);
  const vx = bx - ax, vy = by - ay;
  const mag = Math.hypot(vx, vy);
  if (mag === 0) return NaN;
  const cos = Math.max(-1, Math.min(1, Math.abs(vy) / mag));
  return (Math.acos(cos) * 180) / Math.PI;
}

// Torso lean from vertical, signed: + = forward (toward the catch), - = back.
export function signedLean(hip: KP, shoulder: KP, facing: Facing): number {
  const [hx, hy] = up(hip), [sx, sy] = up(shoulder);
  const vx = sx - hx, vy = sy - hy;
  const mag = Math.hypot(vx, vy);
  if (mag === 0) return NaN;
  const deg = (Math.acos(Math.max(-1, Math.min(1, vy / mag))) * 180) / Math.PI;
  const forward = facing === "right" ? vx >= 0 : vx <= 0;
  return forward ? deg : -deg;
}

// Decide facing from averaged nose-vs-hip horizontal offset (confidence-weighted).
export function detectFacing(frames: KP[][]): Facing {
  let num = 0, den = 0;
  for (const f of frames) {
    const nose = f[KEYPOINT.nose];
    const lh = f[KEYPOINT.leftHip], rh = f[KEYPOINT.rightHip];
    if (!nose || !lh || !rh) continue;
    const w = (lh.score + rh.score) / 2;
    const hipX = (lh.x + rh.x) / 2;
    num += (nose.x - hipX) * w;
    den += w;
  }
  if (den === 0) return "right";
  return num / den >= 0 ? "right" : "left";
}
